File: 4.02/Teigha.NET_vc11_amd64dll/Teigha.NET/Examples/OdaDgnAppSwigMgd/Form1.cs
/////////////////////////////////////////////////////////////////////////////// 
// Copyright (C) 2002-2016, Open Design Alliance (the "Alliance"). 
// All rights reserved. 
// 
// This software and its documentation and related materials are owned by 
// the Alliance. The software may only be incorporated into application 
// programs owned by members of the Alliance, subject to a signed 
// Membership Agreement and Supplemental Software License Agreement with the
// Alliance. The structure and organization of this software are the valuable  
// trade secrets of the Alliance and its suppliers. The software is also 
// protected by copyright law and international treaty provisions. Application  
// programs incorporating this software must include the following statement 
// with their copyright notices:
//   
//   This application incorporates Teigha(R) software pursuant to a license 
//   agreement with Open Design Alliance.
//   Teigha(R) Copyright (C) 2002-2016 by Open Design Alliance. 
//   All rights reserved.
//
// By use of this software, its documentation or related materials, you 
// acknowledge and accept the above terms.
///////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Teigha.Core;
using Teigha.TG;

namespace OdaDgnAppMgd
{
  public partial class Form1 : Form
  {
    OdDgHostAppServices _hostApp;
    OdRxSystemServices _sysSrv;
    OdDgDumper m_dumper;
    OdDgDatabase CurDb;
    MemoryManager mMan = null;
    MemoryTransaction mStartTrans = null;
    public Form1()
    {
        mMan = MemoryManager.GetMemoryManager();
        mStartTrans = mMan.StartTransaction();
      #region initialization
      InitializeComponent();
      _sysSrv = new OdExDgnSystemServices();
      Teigha.Core.Globals.odrxInitialize(_sysSrv);
      Teigha.Core.Globals.odgsInitialize();
      Teigha.Core.Globals.odrxDynamicLinker().loadModule("TG_Db");
      Teigha.Core.Globals.odrxDynamicLinker().loadModule("TD_Db.dll", false);
      
      // mind derived Srv is used as we need to override gsBitmapDevice method
      //_hostApp = new OdExDgnHostAppServices();
      _hostApp = new Srv();
      m_dumper = new OdDgDumper();
      #endregion
    }
    private void exitToolStripMenuItem_Click(object sender, EventArgs e)
    {
        mMan.StopTransaction(mStartTrans);
        mMan.StopAll();
      this.Close();
    }

    private void openToolStripMenuItem1_Click(object sender, EventArgs e)
    {
      if (openDgnDialog.ShowDialog() == DialogResult.OK)
      {
        try
        {
          OdDgDatabase db = _hostApp.readFile(openDgnDialog.FileName);
          Tree t = new Tree(db);
          t.FormClosed += new FormClosedEventHandler(TreeFormClosed);
          t.MdiParent = this;
          t.Show();
          t.FillTree();
          CurDb = db;
          //exportSVGToolStripMenuItem.Enabled = true;
          UpdateMainMenu(true);
        }
        catch (Exception ex)
        {
          MessageBox.Show(ex.Message);
        }
      }
    }

    private void TreeFormClosed(object sender, FormClosedEventArgs e)
    {
      UpdateMainMenu(false);
    }

    private void UpdateMainMenu(bool flag)
    {
      toolStripSeparator1.Visible = flag;
      ExportSVG.Visible = flag;
      ExportPDF.Visible = flag;
      ExportDWF.Visible = flag;
      ExportSTL.Visible = flag;
    }

    public void setStatus(string s)
    {
      this.toolStripStatusLabel1.Text = s;
    }
    public VectorizeForm createVectorizationWindow(Tree t, OdDgElementId vectorizedViewId, OdDgElementId vectorizedModelId)
    {
      VectorizeForm v = new VectorizeForm(t, vectorizedViewId, vectorizedModelId);
      v.MdiParent = this;
      v.Show();
      return v;
    }

    // It is simply an example of deriving classes. Nothing more.
    public class OdGiContextForDgDatabaseToExportMine : OdGiContextForDgDatabaseToExport
    {
      public override bool getWeight( uint weight, out uint resultWeight, OdDgLevelTableRecord level)
      {
          resultWeight = 2;
          return false;
      }
    }
        
    private void ExportSVG_Click(object sender, EventArgs e)
    {
        MemoryTransaction mTr = mMan.StartTransaction();
      OdDgDatabase db = CurDb;
      OdDgViewGroupTable table = db.getViewGroupTable();
      OdDgElementIterator it = table.createIterator();
      SaveFileDialog saveSvgDialog = new SaveFileDialog();
      saveSvgDialog.Filter = "SVG files|*.svg";

      for (; !it.done(); it.step())
      {
        OdDgViewGroup viewGroup = OdDgViewGroup.cast(it.item().openObject(OpenMode.kForWrite));
        {
          OdDgElementIterator it1 = viewGroup.createIterator();
          for (; !it1.done(); it1.step())
          {
            OdDgView view = OdDgView.cast(it1.item().openObject(OpenMode.kForWrite));
            bool flag = view.getVisibleFlag();
            if (flag)
            {
              if (saveSvgDialog.ShowDialog() != DialogResult.OK)
              {
                return;
              }
              OdGsModule pModule = (OdGsModule)Teigha.Core.Globals.odrxDynamicLinker().loadModule("TD_SvgExport");
              if (null == pModule)
              {
                MessageBox.Show("TD_SvgExport.tx is missing");
              }
              else
              {
                String fileName = saveSvgDialog.FileName;
                OdStreamBuf file = Globals.odrxSystemServices().createFile(fileName, FileAccessMode.kFileWrite, FileShareMode.kShareDenyNo, FileCreationDisposition.kCreateAlways);
                if (null != file)
                {
                  OdGsDevice dev = pModule.createDevice();
                  if (null != dev)
                  {
                    dev.properties().putAt("Output", file);
                    // size of pixel in device units
                    dev.properties().putAt("LineWeightScale", new OdRxVariantValue(1.0));//1.
                    // size of pixel in device units
                    dev.properties().putAt("Precision", new OdRxVariantValue(9));
                    // where copy images
                    String s_base = saveSvgDialog.FileName;
                    s_base = s_base.Substring(0, s_base.LastIndexOf("\\"));
                    dev.properties().putAt("ImageBase", new OdRxVariantValue(s_base));
                    // prefix to prepend to image name
                    dev.properties().putAt("ImageUrl", new OdRxVariantValue("./"));
                    // default image format
                    dev.properties().putAt("DefaultImageExt", new OdRxVariantValue(".png"));
                    OdDgModel pModel = (OdDgModel)db.getActiveModelId().safeOpenObject();
                    dev.properties().putAt("BackGround", new OdRxVariantValue(pModel.getBackground()));
                    ////////////////////
                    // Set active palette
                    uint[] refClr = OdDgColorTable.currentPalette(db);
                    //bool bCorrected = OdDgColorTable.correctPaletteForWhiteBackground(refClr);
                    refClr[255] = pModel.getBackground();
                    dev.setLogicalPalette(refClr, 256);
                    dev.setBackgroundColor(refClr[255]);
                    
                    // Prepare database context for device
                    OdGiContextForDgDatabase pDgnContext = OdGiContextForDgDatabaseToExport.createObject();
                    pDgnContext.setDatabase(db);
                    // do not render paper
                    pDgnContext.setPlotGeneration(true);
                    // Prepare the device to render the active layout in this database.
                    OdDbBaseDatabasePE pDbPE = OdDbBaseDatabasePE.cast(db);
                    OdGsDevice wrapper = pDbPE.setupActiveLayoutViews(dev, pDgnContext);
                    // Setup device coordinate space
                    wrapper.onSize(new OdGsDCRect(0, 1024, 768, 0));
                    // Initiate rendering.
                    wrapper.update();
                  }
                }
              }
            }
          }
        }
      }
      mMan.StopTransaction(mTr);
    }

    private void fillLayoutList(OdStringArray aLayouts, bool bAllLayouts)
    {
      if (bAllLayouts == true)
      {
        OdDgModelTable pModelTable = CurDb.getModelTable();
        if (pModelTable != null)
        {
          OdDgElementIterator pIter = pModelTable.createIterator();
          for ( ; !pIter.done(); pIter.step() )
          {
            OdDgModel pModel = OdDgModel.cast( pIter.item().openObject() );
            if (null != pModel)
            {
              aLayouts.Add(pModel.getName());
            }
          }
        }
      }  
    }
    private void ExportPDF_Click(object sender, EventArgs e)
    {
        MemoryTransaction mTr = mMan.StartTransaction();
        OdRxModule mod = Globals.odrxDynamicLinker().loadApp("TD_PdfExport");
      if (null == mod)
      {
        MessageBox.Show("Failed to load PDF export module", "Error");
      }

      Export2PdfForm dlgPdf = new Export2PdfForm();
      //using (PDFExportParams Params = new PDFExportParams())
      PDFExportParams Params = new PDFExportParams();
      {
        OdGsPageParams pParams = new OdGsPageParams();
        dlgPdf.m_dPaperH = pParams.getPaperHeight();
        dlgPdf.m_dPaperW = pParams.getPaperWidth();
        if (DialogResult.OK != dlgPdf.ShowDialog())
        {
          return;
        }
        if (!dlgPdf.m_bEnableLayers)
          dlgPdf.m_bExportOffLayers = false;

        OdStringArray arrLayouts = new OdStringArray();
        fillLayoutList(arrLayouts, dlgPdf.m_Layouts);
        Params.setLayouts(arrLayouts);

        OdGsPageParamsArray ppArr = Params.pageParams();
        uint len = (uint)arrLayouts.Count;
        if (len == 0) len = 1;
        ppArr.resize(len);
        Params.setPageParams(ppArr);

        bool bV15 = dlgPdf.m_bEnableLayers || dlgPdf.m_bExportOffLayers;

        Params.setDatabase(CurDb);
        Params.setVersion((PDFExportParams.PDFExportVersions)(bV15 ? PDFExportParams.PDFExportVersions.kPDFv1_5 : PDFExportParams.PDFExportVersions.kPDFv1_4));

        //using (OdStreamBuf file = Globals.odrxSystemServices().createFile(dlgPdf.m_FileName, FileAccessMode.kFileWrite, FileShareMode.kShareDenyNo, FileCreationDisposition.kCreateAlways))
        OdStreamBuf file = Globals.odrxSystemServices().createFile(dlgPdf.m_FileName, FileAccessMode.kFileWrite, FileShareMode.kShareDenyNo, FileCreationDisposition.kCreateAlways);
        {
          OdGiContextForDgDatabaseToExportMine pCtx = new OdGiContextForDgDatabaseToExportMine();
          pCtx.setDatabase(CurDb);
          Params.setTGSpecific(true, pCtx);
          OdGiDefaultContext ctx = Params.giContext();
          Params.setOutput(file);
          Params.setExportFlags(
                         (dlgPdf.m_bEmbedded ? PDFExportParams.PDFExportFlags.kEmbededTTF : 0) |
                         (dlgPdf.m_bSHXAsGeometry ? PDFExportParams.PDFExportFlags.kSHXTextAsGeometry : 0) |
                         (dlgPdf.m_bTTFAsGeometry ? PDFExportParams.PDFExportFlags.kTTFTextAsGeometry : 0) |
                         (dlgPdf.m_bSimpleGeomOpt ? PDFExportParams.PDFExportFlags.kSimpleGeomOptimization : 0) |
                         (dlgPdf.m_bZoomToExtents ? PDFExportParams.PDFExportFlags.kZoomToExtentsMode : 0) |
                         (dlgPdf.m_bEnableLayers ? PDFExportParams.PDFExportFlags.kEnableLayers : 0) |
                         (dlgPdf.m_bExportOffLayers ? PDFExportParams.PDFExportFlags.kIncludeOffLayers : 0) |
                            (dlgPdf.m_bUseHLR ? PDFExportParams.PDFExportFlags.kUseHLR : 0) |
                            (dlgPdf.m_bEncoded ? PDFExportParams.PDFExportFlags.kFlateCompression : 0) |
                            (dlgPdf.m_bEncoded ? PDFExportParams.PDFExportFlags.kASCIIHexEncoding : 0) |
                            (dlgPdf.m_bUseHLR ? PDFExportParams.PDFExportFlags.kExportHyperlinks : 0));


          Params.setTitle(dlgPdf.m_Title);
          Params.setAuthor(dlgPdf.m_Author);
          Params.setSubject(dlgPdf.m_Subject);
          Params.setKeywords(dlgPdf.m_Keywords);
          Params.setCreator(dlgPdf.m_Creator);
          Params.setProducer(dlgPdf.m_Producer);

          UInt32[] CurPalette = OdDgColorTable.currentPalette(CurDb);
          CurPalette[255] = 0x00ffffff; // the same as ODRGB(255, 255, 255); in the similar C++ code extract
          OdDgColorTable.correctPaletteForWhiteBackground(CurPalette);

          Params.Palette = CurPalette;

          PdfExportModule module = new PdfExportModule(OdRxModule.getCPtr(mod).Handle, false); //= PdfExportModule.cast();
          OdPdfExport exporter = module.create();
          UInt32 errCode = exporter.exportPdf(Params);

          if (errCode != 0)
          {
            String errMes = exporter.exportPdfErrorCode(errCode);
            String str;
            str = string.Format("Error number : {0}. \n {1}", errCode, errMes);

            if (errCode == 0x10008)
            {
              str += "\nPlease enable Zoom to extents check box or\ndefine page parameters for layout in page setup dialog.";
            }

            MessageBox.Show("PDF error", str);
          }
        }
      }
      mMan.StopTransaction(mTr);
    }

    private void zoomExtents_Click(object sender, EventArgs e)
    {
      VectorizeForm act = (VectorizeForm)this.ActiveMdiChild;
      act.ZoomExtents();
    }
    public void UpdateZoomExtFlag(bool flag)
    {
      zoomExtents.Visible = flag;
    }

    private void ExportDWF_Click(object sender, EventArgs e)
    {
      MemoryTransaction mTr = mMan.StartTransaction();
      OdRxModule mod = Globals.odrxDynamicLinker().loadApp("TD_Dwf7Export");
      if (null == mod)
      {
          MessageBox.Show("Failed to load DWF export module", "Error");
          return;
      }
      DwfExportModule dwf_mod = new DwfExportModule(OdRxModule.getCPtr(mod).Handle, false);
      
      SaveFileDialog dwfDlg = new SaveFileDialog();
      dwfDlg.Title = "Export to DWF";
      dwfDlg.Filter = "Binary DWF v6.0|*.dwf|Zipped Stream DWF v6.0|*.dwf|Compressed DWF v5.5|*.dwf|Binary DWF v5.5|*dwf|Ascii DWF v5.5|*.dwf|Compressed DWF v4.2|*.dwf|Binary DWF v4.2|*dwf|Ascii DWF v4.2|*.dwf|XPS|*.dwfx";
      if (DialogResult.OK != dwfDlg.ShowDialog())
      {
        return;
      }
///////////////////////////
    DwExportParams Params = new DwExportParams();
    OdGiContextForDgDatabaseToExportMine pCtx = new OdGiContextForDgDatabaseToExportMine();
    //OdGiContextForDgDatabaseToExport pCtx = OdGiContextForDgDatabaseToExport.createObject();
    uint iweight = 0;
    bool dRet = pCtx.getWeight(4, out iweight, null);

    pCtx.setDatabase(CurDb);
    Params.setDatabase(pCtx.database());
    Params.setDwfFileName(dwfDlg.FileName);

    // MKU 03\01\02 - resolved "white background issue"
    UInt32[] refColors = OdDgColorTable.currentPalette(CurDb);
    OdDgModel pModel = (OdDgModel)CurDb.getActiveModelId().safeOpenObject();
    UInt32 background = pModel.getBackground();
    refColors[255] = background;

    // This method should be called to resolve "white background issue" before setting device palette
    //bool bCorrected = OdDgColorTable.correctPaletteForWhiteBackground(refColors);
    Params.Palette = refColors;
    Params.setBackground(refColors[255]);

    Params.setInkedArea(false);                                                        // MKU 1/21/2004
    Params.setColorMapOptimize(false);                                                 // MKU 1/21/2004
    Params.setExportInvisibleText(true);
    //params.bForceInitialViewToExtents = true;

    DwfPageData pageData = new DwfPageData();
        //OdDgElementId pId = m_pDb->getActiveModelId();
        //OdDgModelPtr pModel = OdDgModel::cast(pId.openObject());
    DwfPageDataArray pdArray = new DwfPageDataArray();
    switch(dwfDlg.FilterIndex)
    {
      case 1:
        Params.setFormat(DW_FORMAT.DW_UNCOMPRESSED_BINARY);
        Params.setVersion(DwfVersion.nDwf_v60);
        
        pageData.sLayout = pModel.getName();//m_pDb->findActiveLayout(true);
        pdArray.Add(pageData);
        break;
      case 2:
        Params.setFormat(DW_FORMAT.DW_ASCII);
        Params.setVersion(DwfVersion.nDwf_v60);
        pageData.sLayout = pModel.getName();//m_pDb->findActiveLayout(true);
        pdArray.Add(pageData);
        break;
      case 3:
        Params.setFormat(DW_FORMAT.DW_COMPRESSED_BINARY);
        Params.setVersion(DwfVersion.nDwf_v55);
        break;
      case 4:
        Params.setFormat(DW_FORMAT.DW_UNCOMPRESSED_BINARY);
        Params.setVersion(DwfVersion.nDwf_v55);
        break;
      case 5:
        Params.setFormat(DW_FORMAT.DW_ASCII);
        Params.setVersion(DwfVersion.nDwf_v55);
        break;
      case 6:
        Params.setFormat(DW_FORMAT.DW_COMPRESSED_BINARY);
        Params.setVersion(DwfVersion.nDwf_v42);
        break;
      case 7:
        Params.setFormat(DW_FORMAT.DW_UNCOMPRESSED_BINARY);
        Params.setVersion(DwfVersion.nDwf_v42);
        break;
      case 8:
        Params.setFormat(DW_FORMAT.DW_ASCII);
        Params.setVersion(DwfVersion.nDwf_v42);
        break;
      case 9:
        Params.setFormat(DW_FORMAT.DW_XPS);
        Params.setVersion((DwfVersion)602);
        pageData.sLayout = pModel.getName();//m_pDb->findActiveLayout(true);
        pdArray.Add(pageData);
        break;
      default:
        //ASSERT(0);
        return;
    }
    Params.setPageData(pdArray);
    OdDwfExport dwf_exp = dwf_mod.create();
    dwf_exp.exportDwf(Params);
    mMan.StopTransaction(mTr);
///////////////////////////
    }

    private void ExportSTL_Click(object sender, EventArgs e)
    {
        MemoryTransaction mTr = mMan.StartTransaction();
        Export2STLDlg dlg = new Export2STLDlg();
        if (DialogResult.OK == dlg.ShowDialog())
        {
            OdRxModule tmp_mod = Globals.odrxDynamicLinker().loadModule("TD_STLExport");
            STLModule module = new STLModule(OdRxModule.getCPtr(tmp_mod).Handle, false);
            OdDbHandle han = new OdDbHandle(dlg.HandleId);
            OdDgElementId targetId = CurDb.getElementId(han);
            if (targetId.isNull())
            {
                MessageBox.Show("Element not found", "Error");
            }
            else
            {
                OdRxObject pCurObj = targetId.safeOpenObject(OpenMode.kForRead);
                OdGiDrawable pDr = OdGiDrawable.cast(pCurObj);
                if (null == pDr)
                {
                    MessageBox.Show("Not drawable element", "Error");
                }
                else
                {
                    using (OdStreamBuf file = Globals.odrxSystemServices().createFile(dlg.FileName, FileAccessMode.kFileWrite, FileShareMode.kShareDenyNo, FileCreationDisposition.kCreateAlways))
                    {
                        OdGeExtents3d extents = new OdGeExtents3d();
                        // !!! calling an OdGiDrawable method getGeomExrents will give a wrong result
                        // !!! we work with a DGN file, thus we should cast to OdDgElement
                        OdDgElement pElt = OdDgElement.cast(pCurObj);
                        pElt.getGeomExtents(extents);

                        double deviation = extents.minPoint().distanceTo(extents.maxPoint()) / 100;

                        OdResult res = module.exportSTL(CurDb, pDr, file, dlg.Binary == false, deviation);
                        if (res != OdResult.eOk)
                        {
                            OdError tmp_err = new OdError(res);
                            MessageBox.Show(tmp_err.description(), "Error");
                        }
                    }
                }
            }
            //han.Dispose();
            //module.Dispose();
        }
        mMan.StopTransaction(mTr);
    }
  };
  public class Srv : OdExDgnHostAppServices
  {
    // override this method in order to be able to perform DWF export of rasters
    public override OdGsDevice gsBitmapDevice()
    {
      try
      {
        OdGsModule pGsModule = (OdGsModule)Globals.odrxDynamicLinker().loadModule("WinGDI.txv");
        return pGsModule.createBitmapDevice();
      }
      catch(OdError err)
      {
      }
      return base.gsBitmapDevice();
    }
  }
}

File: 4.02/Teigha.NET_vc11_amd64dll/Teigha.NET/Examples/VBDwgAppSwigMgd/Form1.vb
Imports System

Imports Teigha.Core
Imports Teigha.TD


Public Class Form1
    Private sysServ As ExSystemServices
    Private hostApp As ExHostAppServices
    Private pDb As OdDbDatabase
    Private helperDevice As OdGsLayoutHelper

    Private Views As New List(Of DisplayControl)()
    Private viewCtrl As DisplayControl

    Public Sub New()

        ' This call is required by the designer.
        InitializeComponent()

        sysServ = New ExSystemServices()
        hostApp = New ExHostAppServices()
        TD_Db.odInitialize(sysServ)
        Globals.odgsInitialize()
        ' Add any initialization after the InitializeComponent() call.
    End Sub
    Private Sub OpenToolStripMenuItem_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OpenToolStripMenuItem.Click
        Dim openDlg As New OpenFileDialog()
        openDlg.Filter = "dwg files (*.dwg)|*.dwg"
        If openDlg.ShowDialog() = DialogResult.OK Then
            pDb = hostApp.readFile(openDlg.FileName)
            TableLayoutPanel1.RowCount = 1
            TableLayoutPanel1.ColumnCount = 1
            viewCtrl = New DisplayControl()
            viewCtrl.TDDatabase = pDb
            viewCtrl.ResetDevice()
            viewCtrl.Dock = DockStyle.Fill
            viewCtrl.Margin = New Padding(1)
            TableLayoutPanel1.Controls.Add(viewCtrl)
            ExportToPdf.Enabled = True
        End If

    End Sub
    Private Sub FillLayouts(ByRef aLayouts As OdStringArray, ByVal bAllLayouts As Boolean)
        If True = bAllLayouts Then
            Dim pLayoutDict As OdDbDictionary = pDb.getLayoutDictionaryId().safeOpenObject()
            Dim pIter As OdDbDictionaryIterator = pLayoutDict.newIterator()
            While False = pIter.done()
                Dim id As OdDbObjectId = pIter.objectId()
                Dim pLayout As OdDbLayout = id.safeOpenObject()
                If Nothing Is pLayout Then
                    Continue While
                End If

                pIter.next()
            End While
        End If
        If False = bAllLayouts Then

        End If
    End Sub

    Private Sub ExportToPdf_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ExportToPdf.Click
        Dim dlgPdf As New PDFParams()
        Dim Params As New PDFExportParams()
        Dim pParams As New OdGsPageParams()
        'dlgPdf.m_dPaperH = pParams.getPaperHeight()
        'dlgPdf.m_dPaperW = pParams.getPaperWidth()
        If Not dlgPdf.ShowDialog() = DialogResult.OK Then
            Return
        End If

        If Not dlgPdf.m_bEnableLayers Then
            dlgPdf.m_bExportOffLayers = False
        End If

        Dim arrLayouts As New OdStringArray()
        FillLayouts(arrLayouts, dlgPdf.m_Layouts)
        Params.setLayouts(arrLayouts)

        Dim ppArr As OdGsPageParamsArray
        ppArr = Params.pageParams()
        Dim len As UInt32
        len = arrLayouts.Count
        If 0 = len Then
            len = 1
        End If
        ppArr.resize(len)
        Params.setPageParams(ppArr)

        Dim bV15 As Boolean
        bV15 = dlgPdf.m_bEnableLayers Or dlgPdf.m_bExportOffLayers

        Params.setDatabase(pDb)
        If bV15 = True Then
            Params.setVersion(PDFExportParams.PDFExportVersions.kPDFv1_5)
        End If
        If bV15 = False Then
            Params.setVersion(PDFExportParams.PDFExportVersions.kPDFv1_4)
        End If

        Dim file As OdStreamBuf
        file = Globals.odrxSystemServices().createFile(dlgPdf.FileName, FileAccessMode.kFileWrite, FileShareMode.kShareDenyNo, FileCreationDisposition.kCreateAlways)
        Params.setOutput(file)
        Dim flags As PDFExportParams.PDFExportFlags
        flags = 0
        If dlgPdf.m_bEmbedded = True Then
            flags = PDFExportParams.PDFExportFlags.kEmbededTTF Or flags
        End If
        If dlgPdf.m_bSHXAsGeometry = True Then
            flags = PDFExportParams.PDFExportFlags.kSHXTextAsGeometry Or flags
        End If
        If dlgPdf.m_bTTFAsGeometry = True Then
            flags = PDFExportParams.PDFExportFlags.kTTFTextAsGeometry Or flags
        End If
        If dlgPdf.m_bSimpleGeomOpt = True Then
            flags = PDFExportParams.PDFExportFlags.kSimpleGeomOptimization Or flags
        End If
        If dlgPdf.m_bZoomToExtents = True Then
            flags = PDFExportParams.PDFExportFlags.kZoomToExtentsMode Or flags
        End If
        If dlgPdf.m_bEnableLayers = True Then
            flags = PDFExportParams.PDFExportFlags.kEnableLayers Or flags
        End If
        If dlgPdf.m_bExportOffLayers = True Then
            flags = PDFExportParams.PDFExportFlags.kIncludeOffLayers Or flags
        End If
        If dlgPdf.m_bUseHLR = True Then
            flags = PDFExportParams.PDFExportFlags.kUseHLR Or flags
        End If
        If dlgPdf.m_bEncoded = True Then
            flags = PDFExportParams.PDFExportFlags.kFlateCompression Or flags
        End If
        If dlgPdf.m_bEncoded = True Then
            flags = PDFExportParams.PDFExportFlags.kASCIIHexEncoding Or flags
        End If
        If dlgPdf.m_bExportHyperlinks = True Then
            flags = PDFExportParams.PDFExportFlags.kExportHyperlinks Or flags
        End If
        If dlgPdf.m_bOptimized = True Then
            flags = PDFExportParams.PDFExportFlags.kEmbededOptimizedTTF Or flags
        End If
        If dlgPdf.m_bAsciiHex = True Then
            flags = PDFExportParams.PDFExportFlags.kASCIIHexEncoding Or flags
        End If
        If dlgPdf.m_bLinearize = True Then
            flags = PDFExportParams.PDFExportFlags.kLinearized Or flags
        End If
        '(m_bMergeLinesCrossing && m_iPdfAMode == (unsigned int)PDFExportParams::kPDFA_None ? PDFExportParams::kMergeLines : 0) |
        If dlgPdf.m_bLinearize = True And dlgPdf.pdfaCtrl = 0 Then
            flags = PDFExportParams.PDFExportFlags.kMergeLines Or flags
        End If


        Params.setExportFlags(flags)


        Params.setTitle(dlgPdf.str_Title)
        Params.setAuthor(dlgPdf.str_Author)
        Params.setSubject(dlgPdf.str_Subject)
        Params.setKeywords(dlgPdf.str_Keywords)
        Params.setCreator(dlgPdf.str_Creator)
        Params.setProducer(dlgPdf.str_Producer)

        Params.setUseViewExtents(dlgPdf.m_bIgnoreInvisible)
        Params.setImageCropping(dlgPdf.bCropInvisible)
        Params.setDCTCompression(dlgPdf.bAdditionalCompression)
        Params.setDCTQuality(dlgPdf.imageQ)
        If dlgPdf.pdfaCtrl = 0 Then
            Params.setArchived(PDFExportParams.PDF_A_mode.kPDFA_None)
        End If
        If dlgPdf.pdfaCtrl = 1 Then
            Params.setArchived(PDFExportParams.PDF_A_mode.kPDFA_1b)
        End If

        'for some reason SearchableTextType enum isn't available for now, thus this parameter setting is skipped for now
        ' bug SWIG-472 opened to resolve this issue
        'params.setSearchableTextType(STType);

        If dlgPdf.m_bEnablePRC = True Then
            If dlgPdf.m_bAsBrep = True Then
                Params.setPRCMode(PDFExportParams.PRCSupport.kAsBrep)
            End If
            If dlgPdf.m_bAsBrep = False Then
                Params.setPRCMode(PDFExportParams.PRCSupport.kAsMesh)
            End If

            Dim pModule As OdRxModule = Globals.odrxDynamicLinker().loadApp("OdPrcModule")
            If pModule IsNot Nothing Then
                pModule = Globals.odrxDynamicLinker().loadApp("OdPrcExport")
            End If

            If pModule Is Nothing Then
                Return
            End If

            Dim compressLevel As PDFExportParams.PRCCompressionLevel = PDFExportParams.PRCCompressionLevel.kA3DLooseCompression
            Dim bCompress As Boolean = True
            Select Case dlgPdf.ui_Compression
                Case 0
                    bCompress = False
                Case 2
                    compressLevel = PDFExportParams.PRCCompressionLevel.kA3DMeddiumCompression
                Case 3
                    compressLevel = PDFExportParams.PRCCompressionLevel.kA3DHighCompression
            End Select
            If bCompress = True Then
                Params.setPRCCompression(compressLevel, True, True)
            End If
        End If

        Select Case dlgPdf.ui_SolidHatches
            Case 1 ' as a drawing (vectorizer)
                Params.setSolidHatchesExportType(PDFExportParams.ExportHatchesType.kDrawing)
            Case 2 ' as a pdf path
                Params.setSolidHatchesExportType(PDFExportParams.ExportHatchesType.kPdfPaths)
            Case 0 ' as a bitmap
            Case Else
                Params.setSolidHatchesExportType(PDFExportParams.ExportHatchesType.kBitmap)
        End Select

        Select Case dlgPdf.ui_GradientHatches
            Case 1 ' as a drawing (vectorizer)
                Params.setGradientHatchesExportType(PDFExportParams.ExportHatchesType.kDrawing)
            Case 0 ' as a bitmap
            Case Else
                Params.setGradientHatchesExportType(PDFExportParams.ExportHatchesType.kBitmap)
        End Select

        Select Case dlgPdf.ui_OtherHatches
            Case 1 ' as a drawing (vectorizer)
                Params.setOtherHatchesExportType(PDFExportParams.ExportHatchesType.kDrawing)
            Case 0 ' as a bitmap
            Case Else
                Params.setOtherHatchesExportType(PDFExportParams.ExportHatchesType.kBitmap)
        End Select

        Params.setGeomDPI(dlgPdf.vectorRes)
        Params.setBWImagesDPI(dlgPdf.imagesRes)
        Params.setColorImagesDPI(dlgPdf.imagesRes)

        Dim bMonoPalette As Boolean = False

        ' for some reason PDFExportParams.ColorPolicy considered ambiguous - bug swig 473 opened
        'Select dlgPdf.ui_ColorPolicy
        '    Case 1
        'Params.setColorPolicy(PDFExportParams.ColorPolicy.kMono)
        '    Case 2
        'Params.setColorPolicy(PDFExportParams.ColorPolicy.kGrayscale)
        '    Case 3
        'bMonoPalette = True
        '    Case 0
        '    Case Else
        'End Select

        Dim bMonochrome As Boolean = False
        Dim pValidator As OdDbPlotSettingsValidator

        Dim i As Boolean = pDb.getPSTYLEMODE()
        Dim strMono As String = "monochrome.stb"
        If i = True Then
            strMono = "monochrome.ctb"
        End If

        If bMonoPalette = True  Then
            pValidator = hostApp.plotSettingsValidator()
            Dim PSSlist As OdStringArray = New OdStringArray()
            pValidator.plotStyleSheetList(PSSlist)

            Dim iSize As Integer = PSSlist.Count
            Dim j As Integer = 0
            While j < iSize
                If strMono = PSSlist.Item(j) Then
                    bMonochrome = True
                    j = iSize
                End If
                j = j + 1
            End While
        End If

        If bMonochrome = True Then
            Dim pLayoutDict As OdDbDictionary = pDb.getLayoutDictionaryId().safeOpenObject()
            Dim f As Integer = 0
            Dim lSize As Integer = Params.layouts.Count
            While f < lSize
                Dim pLayout As OdDbLayout = pLayoutDict.getAt(Params.layouts()(f)).safeOpenObject(OpenMode.kForWrite)
                pValidator.setCurrentStyleSheet(OdDbPlotSettings.cast(pLayout), strMono)
                f = f + 1
            End While
        End If
        ''''''''''''''''''''''''''''''''''''
        Dim CurPalette() As UInteger
        CurPalette = Teigha.Core.AllPalettes.getLightPalette()

        Params.Palette = CurPalette
        Params.setBackground(CurPalette(0))

        Params.setHatchDPI(dlgPdf.hatchRes)
        'Params.setGeomDPI(dlgPdf.m_diGeomDPI)

        Dim pdf_mod As OdRxModule = Globals.odrxDynamicLinker().loadApp("TD_PdfExport")
        Dim pdf_module As PdfExportModule = New PdfExportModule(OdRxModule.getCPtr(pdf_mod).Handle, False)
        Dim exporter As OdPdfExport = pdf_module.create()
        Dim errCode As UInt32 = exporter.exportPdf(Params)
    End Sub
End Class
